### DISCUSSION TITLE
How can I generate BBW for 2D?

### DISCUSSION THREAD

**Discussion by orange4glace** [libigl/libigl#1951] [Q&A] [answered]
First, super thanks to such a wonderful library.

I'm trying to generate [BBW for 2D](https://youtu.be/P9fqm8vgdB8?t=99) (specifically bone-edges) but struggling with it since I have a little knowledge about this domain 😭

If anyone could help, any little example code (or what [`V`, `Ele`, `b`, `bc`](https://github.com/libigl/libigl/blob/main/include/igl/bbw.h#L38) means when it comes to 2D) would be much appreciated.

Thanks !

**Comment by orange4glace** [comment]
While researching, I found that BBW only works (at least for bone type) if the position of control point exactly matches with one of the vertex position of the polygon.
I'm curious whether this is intended or not.

**Reply by alecjacobson** [reply]
That's right, the mesh over which BBWs are computed needs to have vertices and edges along the bone segments. You can use igl::copyleft::triangle::triangulate to achieve this.

**Comment by alecjacobson** [comment, marked as answer]
From the bbw.h 
```
 //   V  #V by dim vertex positions
  //   Ele  #Elements by simplex-size list of element indices
  //   b  #b boundary indices into V
  //   bc #b by #W list of boundary values
```

So for 2D you'd have a triangle mesh of your shape with vertices `V` and face indices in `Ele`.  Any vertex whose index is in `b` will get the corresponding value in `bc` as a fixed constraint. I recommend using `igl::boundary_conditions` to collect these.

**Reply by orange4glace** [reply]
Thanks @alecjacobson :>

I have done with replace `T` by `F` in '*tutorial_403*' which is,

```c++
  igl::readMESH(TUTORIAL_SHARED_PATH "/hair2D.mesh",V,T,F);
  U=V;
  igl::readTGF(TUTORIAL_SHARED_PATH "/hair2D.tgf",C,BE);
  // retrieve parents for forward kinematics
  igl::directed_edge_parents(BE,P);

  // List of boundary indices (aka fixed value indices into VV)
  VectorXi b;
  // List of boundary conditions of each weight function
  MatrixXd bc;
  igl::boundary_conditions(V,F,C,VectorXi(),BE,MatrixXi(),b,bc);
//                           ~ <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< HERE

  // compute BBW weights matrix
  igl::BBWData bbw_data;
  // only a few iterations for sake of demo
  bbw_data.active_set_params.max_iter = 8;
  bbw_data.verbosity = 2;
  if(!igl::bbw(V,F,b,bc,bbw_data,W))
//               ~ <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< AND HERE
  {
    return EXIT_FAILURE;
  }
```

**Reply by orange4glace** [reply]
I have one further question. When I print W, the row size is smaller than number of bone controls.
Specifically, the last bone control seems to be missing.


<img src="https://user-images.githubusercontent.com/17820596/149649094-e8c44cda-df05-476d-bc2b-6597923f8d1c.png" width="700" style="max-width: 100%;">

![image](https://user-images.githubusercontent.com/17820596/149648746-8dc3ef61-7aa0-46a9-9238-cd7eb6434ae1.gif)

In the above image, the first row of W seems to contain both weight of the first bone control and the second bone control.
and there're 11 bone controls in total but W has only 8 rows.
(each connected bone group has one missing weights, (2- 1) + (5 - 1) + (4 - 1) == 8)

Am I missing something or the weight of the last vertex is redundant in bone controls?
